src/checkpoint/dispatch/vrt/base.h

```diff
@@ -0,0 +1,94 @@
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#if !defined INCLUDED_CHECKPOINT_DISPATCH_VRT_BASE_H
+#define INCLUDED_CHECKPOINT_DISPATCH_VRT_BASE_H
+
+#include "checkpoint/dispatch/vrt/registry_common.h"
+#include "checkpoint/dispatch/vrt/base_base.h"
+#include "checkpoint/dispatch/vrt/static_dispatch_typeidx.h"
+#include "checkpoint/dispatch/vrt/object_registry.h"
+#include "checkpoint/dispatch/vrt/serializer_registry.h"
+#include "checkpoint/dispatch/vrt/inheritance_assert_helpers.h"
+#include "checkpoint/dispatch/vrt/serialize_instantiator.h"
+
+#define checkpoint_virtual_serialize_base(BASE)                                   \
+  using _CheckpointVirtualSerializerBaseType = BASE;                              \
+  virtual void _checkpointDynamicSerialize(                                       \
+    void* s,                                                                      \
+    ::checkpoint::dispatch::vrt::TypeIdx ser_idx,                                 \
+    ::checkpoint::dispatch::vrt::TypeIdx expected_idx                             \
+  ) {                                                                             \
+    ::checkpoint::dispatch::vrt::assertTypeIdxMatch<BASE>(expected_idx);          \
+    auto dispatcher =                                                             \
+      ::checkpoint::dispatch::vrt::serializer_registry::getObjIdx<BASE>(ser_idx); \
+    dispatcher(s, *static_cast<BASE*>(this));                                     \
+  }                                                                               \
+  virtual ::checkpoint::dispatch::vrt::TypeIdx _checkpointDynamicTypeIndex() {    \
+    return ::checkpoint::dispatch::vrt::DispatchTypeIdx<BASE>::get();             \
+  }                                                                               \
```

Comment:
Do we need the instantiation call in this macro as well? I think it may only complain in a case where a program defines a non-pure base, and no derived classes that would end up instantiating it by recursion.

Reply:
Does that equate to using virtual serialize without any derived classes? The way I see it, if you any derived classes you will recursively reference the base at some point. I guess it's a valid use mode to not have derived classes, maybe using virtual serialize because you plan to add some later or you plan for code extendability. In fact, this might happen with EMPIRE classes that get derived classes in optional components.

Reply:
Added to base